### ISSUE TITLE
Even more forms weirdness, or...

### ISSUE THREAD

**Issue by jdickey** [ageweke/fortitude#34]
...probably *I'm* just having even more of a brain-dead week than usual. (Yes, I have read Issue #15, along with Issue #21 and tried to wrap my head around Commit 9d7ee0b that closed it, for what those are worth.)

Consider the following, in an otherwise thoroughly-vanilla Fortitude-fortified Rails app:

```ruby

# Class encapsulating all page-specific view code for `things/new`.
class Views::Things::New < Views::Base
  needs :thing, flash: {},
                title_content: 'Add New Thing to List of All Things'

  # include SemanticLogger::Loggable
  include Fortitude::Rails::Helpers

  def content
    page_title_content
    form_for(:thing) do |f|
    end
  end

  private

  def page_title_content
    content_for :title, title_content
  end
end # class Views::Things::New
```

Yet when I run a simple test that uses the class, `form_for` is undefined:

```
$ bundle exec ruby -Itest test/views/things/new_test.rb 

# Running tests with run options --seed 10623:

E
Error:
Views::Things::New#test_0001_calls the #content_for helper correctly when rendered:
NoMethodError: undefined method `form_for' for #<Object:0x007fb89a611d90>
    app/views/things/new.rb:14:in `content'
    test/views/things/new_test.rb:22:in `block (2 levels) in <main>'
    test/views/things/new_test.rb:25:in `block (2 levels) in <main>'


Finished tests in 0.471040s, 2.1230 tests/s, 0.0000 assertions/s.


1 tests, 0 assertions, 0 failures, 1 errors, 0 skips
```

Test code:

```ruby

require 'test_helper'

require File.expand_path('app/views/things/new', Rails.root)

describe 'Views::Things::New' do
  let(:described_class) { Views::Things::New }
  let(:flash) { ActionDispatch::Flash::FlashHash.new }
  let(:thing) { FactoryGirl.build_stubbed :thing }
  let(:obj) do
    described_class.new(thing: thing, flash: flash).tap do |ret|
      ret.class_eval do
        attr_reader :cf_attrib, :cf_value

        def ret.content_for(attrib, value)
          @cf_attrib = attrib
          @cf_value = value
        end
      end
    end
  end
  let(:actual) { obj.to_html }

  it 'calls the #content_for helper correctly when rendered' do
    _ = actual # nothing happens until widget is rendered, so...
    expect(obj.cf_attrib).must_equal :title
    expect(obj.cf_value).must_equal 'Add New Thing to List of All Things'
  end
end # describe 'Views::Things::New'
```

Line 14 of the source is, of course, the call to `form_for`, and Line 22 of the test (mentioned in the backtrace) is the assignment to `actual` which renders the widget (which is where you'd expect it to blow up, I guess). I've been beating my head against one variant or another of this for *days*; what am I missing (again)?

**Comment by jdickey** [comment]
Spelunking in the source and specs the last couple of days, I notice a few things.

There are some very similarly-named modules (e.g., `Fortitude::Rails::Helpers` vs `Fortitude::Rails::FortitudeRailsHelpers` without any immediately-visible distinction between the two.

You run your tests by spawning off a separate Rails server (using your apparently completely undocumented `oop_rails_server` Gem); I'm spending the afternoon trying to understand how that works and how you use it. My guess is that you've decided that really is the only way you can make Rails' helpers available to code being exercised by RSpec or equivalent. Is my understanding then correct that the best use of my time for getting tests like the one mentioned at the top of this issue would be spent in grokking (and documenting) `oop_rails_server` (and probably your `fortitude` specs that use it) in sufficient fullness to use it in my own tests? (I notice also that there are no tests for `oop_rails_server`, are there?)

Right, then; commencing herd-level yak shave. :sweat_smile:

**Comment by jdickey** [comment]
Quoting a status update which I just sent my boss, after _finally_ figuring out (a chunk of) How This Stuff Really Works (IIUC):

> OK, I’m seeing a few things reading through his code. The main “magic” happens [here](https://github.com/ageweke/oop_rails_server/blob/master/lib/oop_rails_server/helpers.rb#L217-L227), where he defines an RSpec class method (`uses_rails_with_template`) that is used, e.g., in [this Fortitude spec](https://github.com/ageweke/fortitude/blob/master/spec/rails/complex_helpers_system_spec.rb#L2). What this does is, before each RSpec test suite (​_not_​ every individual ​_test_​, thank `$DEITY`), it generates, bundles, and spins up a guaranteed-new Rails app with the specified Rails and Ruby versions (which each default to the latest installed on the system the tests are being run on), with controller and view specialisations merged in (that’s what he means in `oop_rails_server` when it talks about a “template”). The tests in an individual spec file are run ​_in the context of that running Rails app server_​, which is then taken down after the last test in that spec file completes. Then it’s on to the next spec file; a good working definition of “slow tests”, but it gets the job done. You’re not playing mind games with Rails, wondering whether or not helpers are available, because you’re running ​_within a Rails app_​ that just happens to be designed and configured for testing purposes.
>
> Oh, did I mention he actually builds, bundles and starts ​_two_​ Rails apps each time a spec calls `uses_rails_with_template`? There’s a `:bootstrap` app and a `:primary` app; IIUC, one is the app server under test and the other is a driver app for that. Wow. But it works.
>
> A working MiniTest port of that would give a ​_much better_​ approximation of “slow tests” because, unlike RSpec, MiniTest doesn’t have separate per-test and per-suite setup/teardown; `before` means “run before each and every test” and `after` likewise. So when he runs 722 tests in 2m5s on my system, he’s running through the startup/teardown cycle 20 times (which does not bracket all his tests, just the ones that require Rails). By my count, that’s 142 individual tests.

Is `oop_rails_server` a superlative achievement, under the circumstances? Absolutely. Is it under-documented? Absolutely. Is it over-engineered for what’s needed? Not at all if you’re interested in CI; this lets you do things like run a CI matrix, testing different versions of Ruby and Rails together, which is a Solid Gold-Class Royal Pain in the Privates otherwise. He can guarantee that his Gem works in known fashion with every version of MRI from 1.8 to 2.2.3 (with JRuby et al along for the ride as desired), and he documents testing for every version of Rails from 2.2.3 (released 27 September 2009) forward until current (4.2.5; two weeks old last Thursday).
>
> Documentation, especially of how and why `fortitude` uses `oop_rails_server`, would have been Very Nice (and would have saved me at least a week of wheel-spinning).
>
> But the short answer appears to be no, you can’t do MiniTest-based testing of widgets that exercise Rails helpers in the idiomatic fast-tests MiniTest style. That doesn’t mean your widgets don’t work (especially if they’re simple enough); it doesn’t even mean you can’t test them (though I’ve yet to research any gotchas mixing MiniTest and RSpec tests).
>
> Bloody amazing.

Do I generally have this right, especially the second-last paragraph ("But the short answer&hellip;")?

**Comment by ageweke** [comment]
Jeff,

First off — my apologies for two things: one, how much work you’ve had to do in order to get your head around this, and, two, how I’ve been out of the loop recently. I’m realizing there’s an irony in here: for _users_ of Fortitude, very many of them likely are getting paid by their jobs to do work that happens to touch on Fortitude, and so they can spend work hours working on it — yet, because it’s a side project for me, I have to relegate all Fortitude work to off-hours. Recently, that’s had severe competition from “get re-settled after hiking all summer”. I’m hoping that changes soon.

And I’m sorry that Fortitude has so few docs right now, and so so much of this stuff has to be figured out all over again. That’s exactly why I released it as a beta (< 1.0.0), but I think I’ve also learned an important lesson: docs are probably as critical as anything else, and I shouldn’t leave them to the end. Obviously, I can’t go back and change that now, but I will say that, once I clear out the filed bugs, docs are my highest priority and will remain so until they’re done.

Let me start with the email you quote above: you have deduced how this stuff works almost completely correctly, as well as the motivations for it. I spin up a real, honest-to-God Rails app, from scratch, out of process, exactly because I don’t know any way to do this stuff in-process or in a “faster” way that produces results I can have confidence in — it doesn’t let me hit the full CI matrix, test configuration files, figure out if Fortitude can work both with and without Erector installed, and so on. And, with certain key features (_e.g._, the `--local` flag to `bundle install`), it all actually works plenty fast (IMHO).

Just one clarification — and I think it’s largely immaterial:

* The `:bootstrap` app is actually nothing more than setting up a `Gemfile` that references `rails` and then running `bundle install` against that, so we can `rails` installed and can do `rails new`. It doesn’t actually create a new Rails app on disk, spin up a Rails server, or anything like that.

Having said all this, all of the stuff above is about how I test Fortitude itself, which should be different from “how do I write tests against Fortitude views?”. My goal with the latter is to make it be _exactly the same_ as writing tests against ERb views, HAML views, and so on. I’m going to grab exactly your spec, above, reproduce the error, and figure out what’s wrong. This stuff is supposed to “just work”, and, to the extent it doesn’t, that’s a problem, and I want to fix it. Any bugs I find I’ll include as specs in the Fortitude spec suite, so they never recur.

Finally, this is something I don’t think I’ve articulated clearly before, but that’s really important, and should end up in the Fortitude docs. Here’s the fundamental issue with `form_for` and friends:

ERb’s `<%= ... %>` is exactly equivalent to Fortitude’s `rawtext(…)`, and ERb’s `<% … %>` is exactly equivalent to Fortitude’s `…` — that is, nothing is needed in the latter case. The tricky part is this — here’s how you use `form_for` in ERb:

```erb
<%= form_for(:foo) do |f| %>
  …
  <%= f.text_field … %>
  …
<% end %>
```

Using the mapping above, this would translate to the following Fortitude code:

```ruby
  def content
    …
    rawtext(form_for(:foo) do |f|
      …
      rawtext(f.text_field …)
      …
    )
    …
  end
```

This would work perfectly fine, but it’s messy as hell, and nobody wants to write Fortitude like this. The purpose of `YieldedObjectOutputter` and the `helper :form_for, :transform => :output_return_value` stuff is exactly to say “hey Fortitude, when I call `form_for`, you should go ahead and actually output its return value, instead of making me wrap it in `rawtext()`” (that’s the `:transform => :output_return_value`), and, “inside that block, make sure if I call `#text_field` on the yielded object, you should output that, too, instead of making me wrap it in `rawtext()`” (that’s the `YieldedObjectOutputter`).

There isn’t anything “funny” going on here other than this — if everyone were OK writing all those `rawtext()` calls, we wouldn’t need any of this madness. And, fundamentally, it’s a judgment call: should `link_to` output directly, instead of needing a `rawtext()` around it? (My answer: yes, absolutely.) Should `number_to_currency` output directly, instead of needing a `rawtext()` around it? (My answer: definitely not.) But there’s zero difference in the Rails source base between these two helpers, and so Fortitude has to pick up the slack in order to make writing code easier.

I haven’t yet gotten a chance to dig in to your example yet (I’ve been working on the bug about supporting all Rails view paths, not just `app/views` so far, but getting close to done on that, and the amount of time I’ve been able to dedicate has been minimal — but should increase shortly), so I’m not 100% confident that’s the issue there. And maybe you already know all of that, and this isn’t helping at all. Either way, I’ll pick up your code and run with it. But hopefully this is informative, at least for whoever comes after you, and possibly you; and, either way, this is clearly an important section of the docs to write for Fortitude.

I should be able to get into this stuff this weekend — I’ll let you know much more as soon as I can. My sincerest apologies for all the yak-shaving you’ve had to do! My goal is for this all to “just work”, and that’s what I want to do ASAP — in many ways, that defines what a 1.0.0 release would be, in my eyes (including documentation).

**Comment by jdickey** [comment]
Andrew,

Wow again. Thanks for the detailed reply (and in particular the refresher on ERb, which I haven't touched with malice aforethought in nearly five years now). It's nice to have confirmation that my (somewhat corrected) understanding of things should match reality closely enough to do useful work.

I'll try the `rawtext` wrappers around the `form_for`-related stuff, but I don't know how optimistic I am. If `form_for` were simply blowing up and spewing garbage or an internal error, that would be one thing, but it's simply *not being seen* when my view calls it:

As quoted in the initiating comment:

```
$ bundle exec ruby -Itest test/views/things/new_test.rb 

# Running tests with run options --seed 10623:

E
Error:
Views::Things::New#test_0001_calls the #content_for helper correctly when rendered:
NoMethodError: undefined method `form_for' for #<Object:0x007fb89a611d90>
    app/views/things/new.rb:14:in `content'
    test/views/things/new_test.rb:22:in `block (2 levels) in <main>'
    test/views/things/new_test.rb:25:in `block (2 levels) in <main>'


Finished tests in 0.471040s, 2.1230 tests/s, 0.0000 assertions/s.


1 tests, 0 assertions, 0 failures, 1 errors, 0 skips
```

Or am I misreading the error message, not considering some context that I should be?

**Comment by ageweke** [comment]
Nah, we’re on the same page — it’s pretty darn clear that there’s something different going on in your case, since, as you said, `#form_for` is just not being seen at all. 

I wanted to give the background above exactly because it’s this deep and difficult issue around `form_for` and similar helpers, and (even for me, and I wrote the damn thing) Fortitude’s interactions with it can be really tricky. If you’re running into issues anywhere near it, it’s useful to know.

However, your issue, yeah, is almost certainly something else. I’m going to throw it into my `spec` tree tomorrow and see where I can get with it. I’ve never used MiniTest, but testing frameworks are all vaguely close enough that I’m sure I can figure it out rapidly. (FWIW, I’m actually a much bigger fan of traditional `assert`-style testing frameworks than the insane amount of magic that is RSpec; I only use RSpec because it seems to have gained so overwhelmingly much traction out there that it’s probably worth it. But that doesn’t mean I love it. ;) I grew up in testing on plain ol’ JUnit, and that was just fine…well, except that it was Java…)

**Comment by jdickey** [comment]
&hellip;and no; changing my `#content` method above to

```ruby
  def content
    page_title_content
    rawtext(form_for(:thing) do |f|
    end)
  end
```

yields the same `undefined method` for `form_for` error as before. Or am I misunderstanding again?

**Comment by jdickey** [comment]
Just saw your reply, *after* leaving my own last. And you may well find something working in RSpec that fails spectacularly in MiniTest, simply because RSpec does _so_ much more for/to/with you beneath the sheets. We left it for MiniTest as the value of "explicit is better than implicit" was ground into our faces over the course of a few projects where Mysterious Things happened in test code that were irreproducible elsewhere. I hope your voyages are more successful.

Yeah, I hear you on assert-style tests and JUnit. I went from JUnit to CppUnit to PHPUnit and ran screaming away after 10+ years to Ruby (first RSpec and now MiniTest::Spec). I prefer the spec-style readability, just without the magical pixie dust. ("Any test tool that is powerful enough to give you everything you want is powerful enough to break everything you have", to paraphrase Jefferson.)

**Comment by ageweke** [comment]
OK, so I've been digging into this today. The problem has nothing to do with RSpec, MiniTest, or Fortitude’s own tests of itself — what’s being measured above. Rather, it’s about the “context” in which a view is rendered. 

When Rails renders a view, it creates a `view_context` that it supplies to the templating engine. The view context is an instance of `ActionView::Base`, and, critically, it’s had a few things done to it: any assignments from the controller (_e.g._, `@foo = …`) have been copied into it (via some relatively gross Ruby metaprogramming magic), and helper methods have all been added to it. It's worth noting that _all_ this stuff depends on a controller instance — obviously it does for the data (`@foo`), but it also does for helper methods, since controllers can say things like `helper :bar` and thus include more helper methods that aren’t available in other cases. In other words, in Rails, the way (and only way) you’re supposed to get a `view_context` is from a controller instance.

Most rendering engines (ERb, HAML, _etc._) actually just generate a whole ton of Ruby code from a template. This Ruby code gets run in the context of this `view_context` (via more Ruby metaprogramming magic). This is exactly the mechanism by which you can just call `form_for`, `link_to`, and any other Rails helper from most templating languages, and why you can access `@foo` and have it just work — you’re actually running your code inside this `view_context` object. To my mind, it’s all a little bit messy (this `view_context` feels like a big slurry of “stuff” that Rails throws in a big pot), but it definitely works.

Fortitude is compatible with this exact same mechanism, but, I think, deals with it a little more cleanly. Fortitude has the concept of a _rendering context_ — an instance of `Fortitude::RenderingContext`. When you call `to_html` on a widget, you can pass in a rendering context, and it will use this for its “connecting points” with the outside world — this is how it knows what object to invoke helpers on (whatever you pass to `RenderingContext.new(:helpers_object)`), what to yield to if you call `yield` from a layout (whatever you pass to `RenderingContext.new(:yield_block)`), and so on. In other words, `RenderingContext` encapsulates all the communication between a widget and, well, the context in which it’s rendered.

There’s [code in Fortitude](https://github.com/ageweke/fortitude/blob/554b8c4df90f45ab2be56fdc6832dfc7b6bb4925/lib/fortitude/rails/renderer.rb#L29) (which calls the [important part](https://github.com/ageweke/fortitude/blob/554b8c4df90f45ab2be56fdc6832dfc7b6bb4925/lib/fortitude/rails/rendering_methods.rb#L13-L17)), then, that creates a `RenderingContext` properly from Rails’ supplied `view_context` (via the `template_handler` here) when you render a view from a controller, and that view happens to be a Fortitude widget — it [passes that `RenderingContext` into the widget’s `to_html` method](https://github.com/ageweke/fortitude/blob/554b8c4df90f45ab2be56fdc6832dfc7b6bb4925/lib/fortitude/rails/renderer.rb#L37-L39). (This is actually `render_to`, not `to_html`, but they’re effectively equivalent. `to_html` just returns a string version of its output; we don’t want to do this in Rails every time, because it can generate unnecessary garbage. `to_html` is a [very thin wrapper around `render_to`](https://github.com/ageweke/fortitude/blob/554b8c4df90f45ab2be56fdc6832dfc7b6bb4925/lib/fortitude/widget/rendering.rb#L44-L47).)

One final note: if you call `to_html` on a Fortitude widget _without_ supplying a `RenderingContext`, it will create a new, empty one for you and use it. This means you won’t have access to any data from the controller, any helpers, or anything to `yield` to (if your view is a layout) — but it will work. (It’s a difficult design choice whether this case should simply fail — meaning it’s more work when you want to say “just render this widget, dammit!” — or use a new, empty `RenderingContext` — meaning it’s easier to forget about hooking up instance variables, helper methods, and so on.)


Sooo…that’s nice and all, but how does this intersect your problem? Well, the issue is that, in your code, you’re actually just instantiating a view object (widget) directly, and there’s no controller instance (and thus no `view_context`) present at all. The view doesn’t have any way of knowing what helpers it should use (since the controller instance defines that), nor anything else it’d get from the controller.

This is where it becomes non-Fortitude-specific: if you removed Fortitude from the equation completely, and changed your view so it used ERb to read in a `.html.erb` file and tried to render that, you’d run into the exact same problem. The code to do this with ERb is slightly messier, but it’s the exact same idea: without being hooked up to a controller instance, basically none of Rails _exists_ to a view, because it’s not getting access to all that data via a `view_context`.

The same’s true for HAML, Slim, or any other rendering engine. I think it’s actually just easier to run _into_ with Fortitude because it’s so much easier to ask the engine to render a view into some HTML (`MyWidget.new.to_html`), vs. the more significant and messier gyrations required in ERb, HAML, and other engines. If you start writing that code against ERb or HAML, you rapidly get the feeling that “I must be doing something wrong”, exactly because it’s so messy — it’s just much easier with Fortitude. (I don’t think I should make it _harder_ in Fortitude, though, because if you’re _not_ in a Rails environment, it’s so much easier to render stuff from Fortitude this way.)


As a conclusion: there is no one single “easy” solution to your problem. The answer is something sort of like: you’re trying to render a Rails view without supplying it a Rails controller instance to connect to — and, in Rails, that’s a nonsensical thing to do. Views don’t exist without controller instances. 

If you _did_ have a controller instance, you could pass it in to your view doing something a bit like this:

```ruby
let(:actual) { obj.to_html(Fortitude::RenderingContext.new(:delegate_object => controller.view_context)) }
```

This is all based on something like [Fortitude’s support for `render :widget` in a controller](https://github.com/ageweke/fortitude/blob/554b8c4df90f45ab2be56fdc6832dfc7b6bb4925/lib/fortitude/rails/rendering_methods.rb#L23-L55), which sets up the right context using Rails methods. That still wouldn’t be _totally_ identical to how a real Rails app works, because what you really want to do is let Rails find the template, create the `template_handler` itself, and then invoke Fortitude [just like it does in production usage](https://github.com/ageweke/fortitude/blob/554b8c4df90f45ab2be56fdc6832dfc7b6bb4925/lib/fortitude/rails/template_handler.rb#L6-L20). But it’d be a whole lot closer, and you’d have your `form_for` back.


I’m _completely_ willing to do work on Fortitude to help solve this problem — the issue is that, I _think_, it’s not really in scope. Different testing frameworks allow for testing views in different ways — exactly because of this problem I’ve been describing above. And, from some basic research, they seem to all do it in the context of a controller or integration test, which makes a lot of sense, given what I was saying above: you just kind of have to have a controller instance in order to render a view in a way that makes any kind of sense at all. For example, [this](https://www.safaribooksonline.com/library/view/rails-4-test/9781680500493/f_0080.html) seems to indicate that Minitest tests views as part of a controller or integration test. 

I hope this has been helpful…and not _too_ long-winded. ;) Here’s a thought: if you find a way of testing views that works well with, say, ERb or HAML (or whatever else you want), and you have any trouble getting it adapted to Fortitude at all — or you can get it working, and it’s just messy — I’ll help out immediately. I’m deep enough in some other Fortitude bugs, and the whole realm of testing views seems broad enough, that running off and trying to learn how to test views (which is, IME, something most people usually skip — perhaps wrongly, but it does seem to be the common case out there) doesn’t seem like it makes a lot of sense, and I suspect it’ll “just work” if you use Fortitude in the exact same context you’d normally use ERb or HAML or whatever.

By all means, let me know if you have any further questions, or if any of this doesn’t make any sense. I’m sorry this was such a long journey for you! Hopefully we’re on our way to bringing that journey to a better end. :)

**Comment by jdickey** [comment]
Andrew, many, *many* thanks; this is extremely helpful. Before I came to Fortitude, I was one of *those* Rails devs who said "don't bother writing view specs; use controller and feature specs instead" exactly because it's such a royal pain to get anything useful done well enough that you don't immediately want to take an hour-long shower afterwards. I assumed* (with the gentlest of nudges from the "Views As Code" tagline) that I'd be able to work with Fortitude widgets just as with any other POROs, and I *can* — right up to but not including the point where Rails-specific code becomes involved. That's when the tasty bonbon melting in my mouth reveals itself as an IED — *not* your fault.

With the explanation of Fortitude's (or equivalent's) interfacing to the guts of Rails' "clean separation of concerns" view support, and reading the (extremely illuminating) code bits you linked to, I understand that 'there is no one single "easy" solution to the problem'. Would it be feasible/worthwhile, in your opinion, (obviously having stared at the tonsils of the beast far more than I) to mock the `view_context` (or `RenderingContext`, rather) so that unit tests for Rails-touching widgets are a) practical and b) cleaner than just falling back to the old-school "don't test views; write controller and feature tests instead" approach? For getting _our_ current job done, the latter would seem to be the faster approach, but It Would Be Wonderful If™ we could develop the former (scaffold/mock of necessary context). This would, as you suggest, likely not be core to Fortitude itself, but why not a 'helper' Gem? Does this make any sense at all, or did my brain turn into a pumpkin at midnight again? :stuck_out_tongue_winking_eye: 

Again; thanks for the explanation and food for thought. Feel free to close this issue as well.

**Comment by ageweke** [comment]
Jeff,

It’s extremely easy to plug in a mock here, if you’d like:

```ruby
rc = Fortitude::RenderingContext.new(:helpers_object => my_helpers_mock)
html_text = Views::Things::New.to_html(rc)
```

The `helpers_object` passed to `RenderingContext.new` is what Fortitude calls helper methods on. So, above, if `my_helpers_mock` is an object that responds to `form_for`, then `form_for` will work (using that implementation), and so on. You can, in fact, use any class you want for that `my_helpers_mock`, and do things like `include [ActionView::Helpers::FormHelper](http://api.rubyonrails.org/classes/ActionView/Helpers/FormHelper.html)` in order to get the real Rails `form_for`, too. So then it’s not even an actual mock, but a real implementation.

But what’s the catch, you ask? Of course there’s one — and it’s a big one. All those modules in Rails that provide methods like `#form_for` are tied together in various ways. For example, `form_for` may well, under certain circumstances, expect various other helper modules to be included; some of those are easy to fix (just include them in your object, too), but others are darn near impossible without having all-of-Rails included (for example, `form_for` may want access to your Rails routes, which are obviously per-application and critical). At some point, you end up on a never-ending chase, including more and more stuff into your “mock”, until it eventually gets to be nearly all of Rails…and now you’re arguably in a worse state, because what you’ve built is something that acts very much — but not _exactly_ — like your real application will, and so you may well have tests providing you false information. And, besides, you can run into other lying tests: say your controller forgets to say `helpers :foo` for some special helpers your view is using, yet you `include` that module into your mock object. Now your test passes, but your real view will break.

My opinion? Rails just isn’t designed for doing this — you can try to do it as best you can, but it really wants to have its whole environment available. In the end, as much as it sucks, I’d personally just go down the path of “make sure you only test views in the context of controller or integration tests”, because that’s the only way to make sure the Rails environment surrounding views is truly set up accurately and correctly. Sorry to be the bearer of bad news…but I think that’s the way it is!

Andrew

**Comment by jdickey** [comment]
Thanks, Andrew.

Yeah, I see where you're coming from on the advice, and that's the way we'll go for the foreseeable future of all our Rails projects. I've seen my job for a couple of years now as being at least as much "mitigating Rails" as "developing with Rails" for precisely the reasons you touch on

As a classical OO developer (Smalltalk/Delphi/Eiffel/Ada, later sideswiped by C++ and Java), Rails drives me *nuts*. Yes, it's an improvement over most of what I saw in PHP (where my response when asked to evaluate one project, "where are the seams" (for testing), was the proverbial last straw…but Rails doesn't have too many "straws" left for me. Yes, it's wonderful to spin something *apparently* useful up in no time; yes, the ecosystem/Gem support is practically unparalleled, but my advice to devs considering Ruby as a new language has become "Ruby is a great language; just *avoid Rails.*" We're getting to the point where we're going to start putting that into practice. Until then, we'll test Fortitude-and-Rails-using views the same way we tested Slim-and-Rails, or Haml-and-Rails…and that sucks with a picobar-level vacuum.

Because it's Rails' fault, for reasons that many, many others have elaborated far better than I could, and it's hard to buck that lock-in.
